### ISSUE TITLE
no-unnecessary-condition False positive when { [_: string]: string }

### ISSUE THREAD

**Issue by yelhouti** [typescript-eslint/typescript-eslint#2883]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

```TS
const filters: { [_: string]: string } = {};
console.log(filters.a?.length);
```

**Expected Result**

no error since filters.a can be (is) undefined

**Actual Result**

2:22 - Unnecessary optional chain on a non-nullish value(@typescript-eslint/no-unnecessary-condition)

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `canary` |
| `@typescript-eslint/parser`        | `canary` |
| `TypeScript`                       | `4.1.2` |
| `ESLint`                           | `7.14.1` |
| `node`                             | `v12.16.2` |

**Comment by bradzacher** [comment]
The type is not declared with null or undefined, so as far as typescript and our rule is concerned - the optional chain is unnecessary.

You must either manually union undefined into your value type, or use the `noUncheckedIndexedAccess` compiler option.

**Comment by yelhouti** [comment]
@bradzacher a is not necessarily present, any key can be present, I should be able to coalsec a. I didn't even define filters in my project. So you logix seems wromg to me please let's discuss this.

**Comment by bradzacher** [comment]
`filter.a` is seen by typescript to be of type `string`.
This means our lint rule sees the type as string.
This means as far as our lint rule is concerned, you're optional chaining on a string, which is unnecessary.

This is working as intended.
There is nothing to discuss here.

Please re-read my comment. There are solutions for you in it. Use the compiler option or union undefined into your type.

**Comment by yelhouti** [comment]
@bradleyayers noUncheckedIndexedAccess although solves my problem seems like a hack, the and should be the default behavior in my opinion.
You can expect every string to be present as a filter, I understand how this can be problematic, example for cases where instead of string I had an enum and all values, should be present, but this more an exception than the rule, and maybe in that case a "!" should be used to specify that all are present.

**Comment by bradzacher** [comment]
I don't think you're understanding me.

Our lint rules only know what TypeScript tells us.

TypeScript tells us that `filters.a` is of type `string`.
We have no choice but to believe it. We can do nothing else as this is all of the information that we've got.

So we see a `string` with an optional chain, and we report on that information.

It doesn't matter what your reason for using an indexed object is or what it might be known by you to contain at runtime.
***We can only act on the information that TypeScript gives us***

If you think the compiler option is too heavy handed and too far reaching - then manually union `undefined` into your object that you think keys will be missing.

```ts
const filters: { [_: string]: string | undefined } = {};
console.log(filters.a?.length);
```

Adding an explicit `undefined` means TypeScript now reports the type of `filters.a` to us as `string | undefined`. So we can act on this and *not* report on the optional chain, as it's valid.

To reiterate:
***We can only act on the information that TypeScript gives us***
TypeScript gives us the information that you give it.
Either give TypeScript the information explicitly with `| undefined`, or use the compiler option to give it implicitly.

**Comment by yelhouti** [comment]
@bradzacher thank you for taking the time, and I assure I understood you the first time, but I still think it's wrong. Typescript doesn't say that any and all keys are present, it says any missing other key than the one mentioned are of type string. your are inferring, that me calling .a assume that a exit which is the wrong inference.
a: string indeed means that a is for type string and can't be undefined.
[_: string]: string means I can have other keys  and all of them are of type string, and not all possible keys exist and are of type string.

**Comment by bradzacher** [comment]
> and I assure I understood you the first time

From your response - you don't.

The type of the key is `string`.
The type of the value is `string`.

Our rule has no idea about the ***key*** and does not know what it should or shouldn't be.

I'm specifically talking about the ***value***.
The fact that the key is typed as `string` has no bearing on types that TS returns for the ***value***.

To clarify, here's an example where the value is instead `number`:

```ts
{
    // Example 1

    const filters: { [_: string]: number } = {};

    const a1 = filters.a;    // typeof === number
    const a2 = filters['a']; // typeof === number
}

{
    // Example 2

    const filters: { [_: string]: number | undefined } = {};

    const a1 = filters.a;    // typeof === number | undefined
    const a2 = filters['a']; // typeof === number | undefined
}
```


In the first example (your code) - typescript will tell us that BOTH `filters.a` AND `filters['a']` return the type `number`.

Which means that - as far as the types are concerned - accessing properties on the value with a `?.` would be ***unnecessary***. I.e. `filters.a?.toString()` and `filters['a']?.toString()` would both ***error*** in our lint rule.

In the second example - I have explicitly added `| undefined` to the ***value***.
Now TypeScript reports that the type of BOTH `filters.a` AND `filters['a']` to be `number | undefined`.

Which means that - as far as the types are concerned - accessing properties on the value with a `?.` would be ***necessary***. I.e. `filters.a?.toString()` and `filters['a']?.toString()` would both ***not error*** in our lint rule.

**Comment by yelhouti** [comment]
> Our rule has no idea about the key and does not know what it should or shouldn't be.

IMO It should ! and explained how I think it should use it:
if key is bound infer defined, else infer can be undefined.
if your code uses a simple typeof, maybe typescript should return `| undefined` when the ke type is unbound

**Comment by bradzacher** [comment]
My question is this - did you play around with or research the compiler option I suggested?
I'm guessing you didn't - or else you'd have seen that it works ***EXACTLY AS YOU HAVE DESCRIBED***.

```ts
{
    // Example 1

    const filters: { [_: string]: number } = {};

    const a1 = filters.a;    // typeof === number | undefined
    const a2 = filters['a']; // typeof === number | undefined
}

declare function get<T>(): T;

{
    // Example 2

    const filters: { [_ in 'a' | 'b']: number } = get();

    const a1 = filters.a;    // typeof === number
    const a2 = filters['a']; // typeof === number
}
```
[repl](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true#code/N4KABBYPRWCiAPAhgWwA4BsCmYCMJxIBjAewDsBnAFzADMBLDKrAJwoC4xgwBtAfU7UW9MgHMAupzIBXFACNWYAL5gAvFyUBuApDClKNJLjV1GzNgDokm3TDBUAnmiwlaa1epnzFAHzDSyABMsBjIsQMIIfWowJAAmEwYmVgoeAHIkNPEbO0dnV3dPWQUWMD8A4NDwkCUCYKIMJBYcWgCiKnpyMFEsKgAeABUAPgAKAEpOAe0QUFtYRFRMHDidYnIYpPMOLl4+MBEwDLSyw7ksqWLFFXUeqnHp3WjDY3VNlKsbSFynFzcPIu8LEienWhgSrzMKXSmWy0FgeV+hTAXhKNSAA)

**Comment by yelhouti** [comment]
@radarsu I tried to work with it, didn't manage to print typeof using console.log...
Understood your point, answered according to it.
Do you agree though? maybe typescipt should return string | undefined in case of unbound key?

**Comment by bradzacher** [comment]
> maybe typescipt should return string | undefined in case of unbound key

It literally does exactly this with the compiler option turned on, as shown in the above example.

If the key is of type `string` - then accessing anything on this

> didn't manage to print typeof using console.log

I don't print using `typeof` - I used intellisense to see the types that TS reports.
See the repl link below each of the examples I posted which will open the TS playground.

**Comment by yelhouti** [comment]
@bradzacher don't you think that should be default?

**Comment by bradzacher** [comment]
Yes, but that would be a ***major*** breaking change that would make it impossible for most codebases to upgrade to that version.

The compilerOption is there to allow codebases to opt-in to the functionality.
Just like the `strict` compiler options - they should be the default, but they're not.

That's got nothing to do with this project though! That's the [TypeScript project's](https://github.com/microsoft/TypeScript/) domain.

**Comment by yelhouti** [comment]
@bradzacher you are completely right, but I didn't know you where using typeof, once I understood that, I understood your frustration with me. Not all of us are familiar with the code base :D
